### ISSUE TITLE
Why fp16 tensor memory usage is larger than fp32 tensor?

### ISSUE THREAD

**Issue by wikiwen** [pytorch/pytorch#97069]
### 🐛 Describe the bug

I found the fp16 tensor memory usage is bigger than fp32 tensor with same number of elements. Exactly, 3 times than it should be! why?
Is it pytorch allocation fp16 tensor by allocating a same fp32 tensor at first and cast to fp16 then? How to reduce this memory usage?

```python
import torch

# Test fp32 tensor
tmp_tensor = torch.empty(1, dtype=torch.int8, device='cuda')
# memory usage: 1115MB (context)
tmp_tensor = torch.empty(1024*1024*1024, dtype=torch.float32, device='cuda')
# memory usage: 5213MB (increase about 4GB, right with tmp_tensor)

# Test fp16 tensor
del tmp_tensor
torch.cuda.empty_cache()
# memory usage: 115MB
tmp_tensor = torch.empty(1024*1024*1024, dtype=torch.float16, device='cuda')
# memory usage: 7621MB (increate about 6GB, wrong with the number tmp_tensor should be)
```

### Versions

PyTorch version: 1.12.1+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A
OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2
Libc version: glibc-2.10
Python version: 3.7.10 | packaged by conda-forge | (default, Oct 13 2021, 21:01:18)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-3.10.0-862.14.1.5.h428.eulerosv2r7.x86_64-x86_64-with-debian-buster-sid
Is CUDA available: True
CUDA runtime version: 10.2.89
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: Tesla V100-PCIE-32GB
Nvidia driver version: 440.33.01
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              8
On-line CPU(s) list: 0-7
Thread(s) per core:  2
Core(s) per socket:  4
Socket(s):           1
NUMA node(s):        1
Vendor ID:           GenuineIntel
CPU family:          6
Model:               85
Model name:          Intel(R) Xeon(R) Gold 6151 CPU @ 3.00GHz
Stepping:            4
CPU MHz:             3000.000
BogoMIPS:            6000.00
Hypervisor vendor:   KVM
Virtualization type: full
L1d cache:           32K
L1i cache:           32K
L2 cache:            1024K
L3 cache:            25344K
NUMA node0 CPU(s):   0-7
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc eagerfpu pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 arat md_clear spec_ctrl intel_stibp flush_l1d
Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.6
[pip3] pytorch-lightning==1.4.2
[pip3] pytorch-metric-learning==1.6.3
[pip3] torch==1.12.1
[pip3] torchmetrics==0.6.0
[pip3] torchtext==0.5.0
[pip3] torchvision==0.13.1
[conda] numpy                     1.21.6                   pypi_0    pypi

**Comment by msaroufim** [comment]
I just tried reproing your results on google colab and couldn't do it - might be some issue with the tool you used to instrument memory usage

![image](https://user-images.githubusercontent.com/3282513/226077689-da19609d-ca29-4857-9b99-845b8d256770.png)

**Comment by wikiwen** [comment]
> I just tried reproing your results on google colab and couldn't do it - might be some issue with the tool you used to instrument memory usage https://colab.research.google.com/drive/1oTJFeCP0H-9myZebyld1XbYoWoD7bUBx?usp=sharing
> 
> ![image](https://user-images.githubusercontent.com/3282513/226077719-069967d1-ef74-4d3a-843d-eaacbdaa3baa.png)

hello, msaroufim, thanks for your reply.  I test your google colab environment. No problem was found in original installed torch1.13.1 and my version torch1.12.1. 

**After more experiments, I guess the reason is my environment: the pycham python console.** As the pycharm python console is able to inspect variable value, may be the fp16 cuda tensor will be cast to fp32 and move to cpu. So the total memory will be 3 times than it should be. I test in common python terminal, the problem was gone.

If the true reason is above, is there some methods to reduce unnecessary memory allocation when inspect variable in pycharm python console or debugger.
